Convert `design-system` package to TypeScript

barklund opened the pull request:
## Context

Conversion of design-system to TypeScript

## Relevant Technical Choices

* Too many to count

## To-do

* [x] Convert something
* [x] Convert the rest
* [x] Fix unit tests
* [x] Fix SVG imports
* [x] Fix lints
* [x] Make sure it actually compiles
* [x] Fix Storybook
* [x] Fix dashboard karma tests
* [x] Fix editor karma tests
* [x] Fix visual inconsistencies as found by Percy

## User-facing changes

_None(?)_

## Testing Instructions

This PR can be tested by following these steps:

1. Smoke-test the entire plugin, from dashboard to editor.

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12659
Fixes #12251
Fixes #11257

Commit:
WIP: Partial conversion of design-system to TS

Commit:
Converted utils and parts of the context menu

Commit:
Fixed import

Commit:
Converted remaining bits of context menu

barklund commented on packages/design-system/src/typings/svg.d.ts:
```diff
@@ -0,0 +1,4 @@
+declare module '*.svg' {
+  const content: React.FunctionComponent<React.SVGAttributes<SVGElement>>;
+  export default content;
+}
```
@swissspidy, this works partially. It works fine in my editor, but when running the TS workflow, I get these errors:

```
packages/design-system/dist-types/icons/index.d.ts:4:40 - error TS2307: Cannot find module './align_bottom.svg' or its corresponding type declarations.

4 export { default as AlignBottom } from './align_bottom.svg';
                                         ~~~~~~~~~~~~~~~~~~~~

packages/design-system/dist-types/icons/index.d.ts:5:40 - error TS2307: Cannot find module './align_center.svg' or its corresponding type declarations.

5 export { default as AlignCenter } from './align_center.svg';
                                         ~~~~~~~~~~~~~~~~~~~~
```

Etc for all 140 SVG's in the design system. Any idea how to make it work correctly?

swissspidy replied:
Does it work with a config like https://github.com/GoogleForCreators/web-stories-wp/blob/main/packages/story-editor/src/typings/svg-imports.d.ts?

barklund replied:
Still only in editor, workflow reports the some problem. Just pushed with this changed typing.

And it's actually not just the workflow that fails. The actual imports fail as well when running the editor. Webpack seems to be missing a loader or can't find the right one:

```
Module parse failed: Unexpected token (1:0)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> <svg width="32" height="32" viewBox="0 0 32 32" fill="none" xmlns="http://www.w3.org/2000/svg">
| <path d="M8.5 21H23.5" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round"/>
| <path fill-rule="evenodd" clip-rule="evenodd" d="M15.5 10.5C14.6716 10.5 14 11.1716 14 12V17C14 17.8284 14.6716 18.5 15.5 18.5H16.5C17.3284 18.5 18 17.8284 18 17V12C18 11.1716 17.3284 10.5 16.5 10.5H15.5Z" fill="currentColor"/>
```

swissspidy replied:
Ah, the loader issue was fixed by #12907, so should work when merging latest changes from main

barklund replied:
Hmm, it compiles with:

```ts
declare module '*.svg' {
  const content: string;
  export default content;
}
```

But then other things break, because they aren't actually string and we need to treat them like React elements (e.g. by adding class names and other properties to them elsewhere in the codebase).

If I do this:

```ts
/**
 * External dependencies
 */
import type { ReactElement, SVGProps } from 'react';

declare module '*.svg' {
  const Icon: (props: SVGProps<SVGElement>) => ReactElement;
  export default Icon;
}
```

It just fails without any errors or warnings (just that no corresponding type can be found). Not really sure why?

I also tried to swap in `SVGAttributes` for `SVGProps` (as it seems more correct), but no luck either.

swissspidy replied:
Hmm then I am not sure either 😕

barklund replied:
@swissspidy, any idea how to resolve this? My only other option, ChatGPT, doesn't know either 😭

barklund replied:
I fixed it! Well, [this answer](https://stackoverflow.com/questions/39040108/import-class-in-definition-file-d-ts) fixed my knowledge:

> Basically, TypeScript has two kind of module types declaration: "local" (normal modules) and ambient (global). The second kind allows to write global modules declaration that are merged with existing modules declaration. What are the differences between this files?
> `d.ts` files are treated as an ambient module declarations only if they don't have any imports. If you provide an import line, it's now treated as a normal module file, not the global one, so augmenting modules definitions doesn't work.

So this works (but stinks):

```ts
declare module '*.svg' {
  const Icon: (
    props: import('react').SVGAttributes<SVGElement>
  ) => import('react').ReactElement;
  export default Icon;
}
```

Alternatively we re-create this declaration in every folder that has SVG's, but there's actually quite a few, as several components have their own icons and graphics, and I'm still not convinced it would work.

So let's just keep this?

barklund replied:
Ah, it's actually as simple as just:

```ts
declare module '*.svg' {
  const Icon: (props: React.SVGAttributes<SVGElement>) => React.ReactElement;
  export default Icon;
}
```

The problem now is, that it doesn't work fully.

Yes, I can build the design system package correctly by running `npx tsc --build` inside the package. It runs correctly (with this file in place) and not without it.

But I cannot build the whole system, because `*.d.ts` files aren't copied to the `dist-types` output (they're only considered necessary **for** the build, not **part of** the build), so once the types have been summarized in dist-types, they're no longer "complete".

If I run `npm run workflow:bundle-packages:types` I get errors like:

```
packages/design-system/dist-types/images/index.d.ts:7:37 - error TS2307: Cannot find module './desert_bw.svg' or its corresponding type declarations.

  export { default as DesertBw } from "./desert_bw.svg";
```

And that makes sense, because once moved to the dist-types folder, then yes, the import no longer has a type for the SVG file. But how do we fix that? I have no clue...

Please halp, @swissspidy, you're my only hope! 🙏

swissspidy replied:
Unfortunately I'm at loss right now as well 😞 

I tried including the definition via `typeRoots` rather than `include`, and the only difference is that the error message is now for the source file and not for the one in `dist-types`. Progress I guess? But to me that sounds like it's not really an issue with copying over `d.ts` file.

The file I used:

```json5
{
  "extends": "../../tsconfig.shared.json",
  "compilerOptions": {
    "rootDir": "src",
    "declarationDir": "dist-types",
    "typeRoots": ["./src/typings", "../../node_modules/@types"]
  },
  "references": [
    { "path": "../i18n" },
    { "path": "../patterns" },
    { "path": "../react" },
    { "path": "../tracking" }
  ],
  "include": [
    "src/components",
    "src/contexts",
    "src/icons",
    "src/images",
    "src/theme",
    "src/types",
    "src/utils"
  ]
}
```

Then I also stumbled upon https://duncanleung.com/typescript-module-declearation-svg-img-assets/ and subsequently tried to move the type definition to a subfolder of `typings`, but to no avail.

I think I need to look at this again later with fresh eyes.

----

Aside: there are a couple of type errors related to `useDebouncedCallback` in `useDropDown` and `useSearch`

barklund replied:
I had an idea about creating a global `typings/` folder at the root and `/// <reference />` that, but it didn't seem to work. Almost, but then not again.

Another idea though: We create a new `svg` package with all the SVG files, and all other packages, that need SVG's import that package. Then we can use whatever hacks we need to make that package work, including perhaps manually crafting a `*.d.ts` if need be. Does that sound stupid?

swissspidy replied:
Can't believe that 6148ff51d589c96d7df5252b5bef189901b6cad8 solves the issue 😭 

The incorrect `declare module` definitions in `story-editor` were breaking this 🤷

barklund replied:
Oh wow, good find! But I thought the `inline-icons` were actually imported as strings as used in various packages? But hey, if it works!

swissspidy replied:
Inline-icons is only used in the moveable package IIRC, so we could move that special case there

Commit:
Fixed some const2enum refs

Commit:
Updated all references to Text and TextSize

Commit:
Updated references to button enums

Commit:
Fixed some configs

Commit:
Attempted fix for svg imports

Commit:
Fixed lints

Commit:
Fixed some tests

Commit:
Fixed more tests

Commit:
Fixed more tests

Commit:
Fixed more tests

Commit:
Fixed remaining component tests

Commit:
Fixed utils test

Commit:
Restored util that was actually in use

Commit:
Fixing exports for dashboard tests to compile

Also fixing a circular dependency

Commit:
Added missing export

Commit:
More export fixes and text references

Commit:
Fixed aria for checkbox

Commit:
Fixed more tests (mostly imports and text refs)

Commit:
Merge branch 'main' into code/12659-ds-ts

# Conflicts:
#	packages/dashboard/src/icons/svg-imports.d.ts
#	packages/design-system/src/contexts/popup/context.js
#	packages/design-system/src/contexts/popup/context.ts
#	packages/design-system/src/contexts/popup/types.ts
#	packages/design-system/src/contexts/popup/usePopup.js
#	packages/design-system/src/contexts/popup/usePopup.ts
#	packages/design-system/src/theme/theme.ts
#	packages/design-system/src/typings/styled.d.ts
#	packages/design-system/src/utils/panelTypes.ts
#	packages/design-system/tsconfig.json

Commit:
Fixed case

Commit:
Fixed some link buttons

Commit:
Trying to fix SVG imports

miina commented on packages/dashboard/src/icons/svg-imports.d.ts:
```diff
@@ -14,7 +14,14 @@
  */
 
+<<<<<<<< HEAD:packages/design-system/src/contexts/popup/types.ts
+export interface PopupState {
```
Merging issues here

barklund replied:
Fixed in 0623bd5bc9

miina commented on packages/design-system/src/utils/useMouseDownOutsideRef.ts:
```diff
@@ -50,8 +49,11 @@ function useMouseDownOutsideRef(onMouseDownOutside = noop) {
     if (node) {
-      handleDocumentClickCaptureRef.current = (evt) => {
-        if (node === evt.target || node.contains(evt.target)) {
+      handleDocumentClickCaptureRef.current = (evt: Event) => {
```
Same here, should these be `MouseEvent`s instead?

barklund replied:
Yeah, it could be. Doesn't change much because we don't use `MouseEvent` specific properties on the event object, but it is a bit more specific.

barklund replied:
Fixed in 3bccfe791f

miina commented on packages/design-system/src/utils/useMouseDownOutsideRef.ts:
```diff
@@ -26,19 +26,18 @@ import { noop } from './noop';
-function useMouseDownOutsideRef(onMouseDownOutside = noop) {
-  const targetRef = useRef(null);
+function useMouseDownOutsideRef<E extends HTMLElement>(
+  onMouseDownOutside: (evt: Event) => void = noop
```
`MouseEvent`?

Commit:
Better signature for util

miina commented on packages/design-system/src/utils/pointerEventsCss.ts:
```diff
@@ -19,7 +19,7 @@
-const pointerEventsCss = css`
+const pointerEventsCss = css<{ pointerEvents?: string }>`
   ${({ pointerEvents }) => {
     if (pointerEvents && typeof pointerEvents === 'string') {
```
Does it still need this check for `string`?

barklund replied:
Hmm, why do we even have this util? It's for adding a single line of CSS with an util which is almost identical to the CSS you would write anyway? We actually even have this stupid util in two different packages! I mean, why? I'll kill both, they're only ever used once and twice respectively!!! It's too stupid to keep as an util!

barklund replied:
Removed in 8a04a9d497.

Commit:
Fixed merge error

miina commented on packages/design-system/src/components/textArea/index.ts:
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright 2021 Google LLC
```
Not related to this PR but this should be 2023 probably? cc @swissspidy

barklund replied:
Yes, I noticed that somewhere by the end. I should update all the new headers, but I mostly just copy-paste it from some other file. Can I automate updating for all new files in this PR, @swissspidy?

swissspidy replied:
Sure

miina commented on packages/design-system/src/components/snackbar/types.ts:
```diff
@@ -0,0 +1,53 @@
+  status?: ThumbnailStatus;
+}
+
+export interface SnackbarNotification {
```
We also have (had?) [`src/types/snackbar.ts`](https://github.com/GoogleForCreators/web-stories-wp/blob/main/packages/design-system/src/types/snackbar.ts).

barklund replied:
Merged in https://github.com/GoogleForCreators/web-stories-wp/pull/12929/commits/87c76d99be8cb6bb12309b4df2ac4242ef15c094.

miina reviewed (commented):
Maybe merge `main` meanwhile?

Commit:
Merged snackbar notification types

Commit:
More specific type

Commit:
Removed pointless utils

miina commented on packages/design-system/src/components/popup/popup.tsx:
```diff
@@ -53,24 +71,31 @@ function Popup({
 
   const positionPopup = useCallback(
-    (evt) => {
+    (evt?: unknown) => {
```
Can it be anything else than `Event` or `undefined`? 🤔

barklund replied:
Can be `Event` from `addEventListener`, `{ width: number; height: number }` from `useResizeEffect`, or `undefined` when invoked directly. We could circumvent the resize effect dimension thingy by wrapping it in another function that just discarded the argument, but I didn't see a strong reason.

miina commented on packages/design-system/src/components/pill/pillGroup.tsx:
```diff
@@ -33,7 +33,19 @@ const Container = styled.div.attrs({
+  extends Omit<ComponentPropsWithoutRef<'button'>, 'onSelect' | 'onClick'> {
+  options: PillOption[];
+  value: string;
+  onSelect: (id: string) => void;
```
No need to add `onClick` back?

barklund replied:
We manually assign an `onClick` handler to the element in question and we don't want a consumer to override that by specifying their own `onClick`, so we just don't allow one to be passed in.

miina commented on packages/design-system/src/components/menu/list/defaultListItem.tsx:
```diff
@@ -37,34 +37,30 @@ const ActiveIcon = styled(CheckmarkSmall)`
-      size={THEME_CONSTANTS.TYPOGRAPHY.PRESET_SIZES.SMALL}
+const DefaultListItem = forwardRef(function DefaultListItem(
+  { option, isSelected, ...rest }: DefaultListItemProps,
+  ref: ForwardedRef<HTMLLIElement>
```
`HTMLLIElement`? Because of `value`?

barklund replied:
Because it's a list item? Not sure I understand the question?

miina commented on packages/design-system/src/components/input/useNumericInput.tsx:
```diff
@@ -52,7 +55,7 @@ export const useNumericInput = ({
    */
   const handleBlur = useCallback(
-    (ev) => {
+    (ev: unknown) => {
```
Why unknown, shouldn't it be an event?

barklund replied:
It is invoked both from a native DOM event listener with a DOM KeyboardEvent (through `useKeyDownEffect`), and through a React event listener (through `onBlur`). So it's either a DOM event or a React event, but we don't actually care, we just pass it on to the `onChange` function (which probably doesn't care either).

It could be typed as `KeyboardEvent | FocusEvent`, which the latter being imported from React, but I found that a bit confusing.

It could also be solved by having a wrapper listener for the React event that invoked it the other with `evt.nativeEvent` ensure it would always receive a proper DOM event, but still, I didn't see a purpose, because we don't actually consume the event.

miina commented on packages/design-system/src/components/datalist/types.ts:
```diff
@@ -0,0 +1,143 @@
+  | { options: O[]; getOptionsByQuery: never }
+  | { options: never; getOptionsByQuery: (query: string) => Promise<O[]> };
+
+export type DataListProps<O extends AbstractOption> =
```
Uh, that's a bunch of props!

barklund replied:
Yeah, it's pretty bloated, because it's used by many different components. However, it is never used directly, I think, only indirectly through other components, so I think it's okay.

miina commented on packages/design-system/src/components/datalist/list/list.tsx:
```diff
@@ -262,44 +267,34 @@ const OptionList = forwardRef(function OptionList(
                     key={option.id || ''}
                     role={'option'}
-                    tabIndex="-1"
+                    tabIndex={-1}
```
Why this change? 🤔

barklund replied:
```
Type 'string' is not assignable to type 'number'.

miina commented on packages/design-system/src/components/datalist/list/list.tsx:
```diff
@@ -262,44 +267,34 @@ const OptionList = forwardRef(function OptionList(
-  listStyleOverrides: PropTypes.array,
-  noMatchesFoundLabel: PropTypes.string,
-};
+const OptionList = memo(forwardRef(OptionListWithRef)) as <
```
🤯

barklund replied:
Yeah, I know. It sucks. I should probably repeat the comment about why this is necessary when having generically typed components.

miina commented on packages/design-system/src/components/contextMenu/contextMenuProvider/provider.tsx:
```diff
@@ -17,23 +17,29 @@
   isHorizontal = false,
   onDismiss,
-}) {
-  const [focusedId, setFocusedId] = useState(-1);
```
Was there a reason for the `-1` before?

barklund replied:
I couldn't find any reference to it at all actually. I think it was just a placeholder that would never match a real id.

miina commented on packages/design-system/src/components/contextMenu/contextMenuProvider/context.ts:
```diff
@@ -0,0 +1,41 @@
+    isHorizontal: false,
+  },
+  actions: {
+    onDismiss: () => undefined,
```
`noop`?

barklund replied:
I actually don't see the benefit of using the `noop` util anywhere. It's not a lot of memory saved, and it's even extra bytes typed importing it. I won't remove it if used, but probably won't use it myself.

miina commented on packages/design-system/src/components/contextMenu/components/subMenuTrigger.tsx:
```diff
@@ -100,20 +112,20 @@ function SubMenuTrigger({
-      pointerTracker.current.x = e.clientX;
-      pointerTracker.current.y = e.clientY;
-      maybeCloseSubMenu();
+    const onPointerMove = (e: Event) => {
```
Should it just be `PointerEvent` directly perhaps?

barklund replied:
This is a native HTML event listener - it is only typed as dispatching a generic event, not typed specifically for each individual event type. Putting `PointerEvent` would result in this error a few lines below.

```
Argument of type '(e: PointerEvent) => void' is not assignable to parameter of type 'EventListenerOrEventListenerObject'.

miina commented on packages/design-system/src/components/banner/index.tsx:
```diff
@@ -0,0 +1,17 @@
+/*
```
^ => `index.ts`

miina reviewed (approved):
Since I'll be off for the next two weeks, approving assuming the remaining issues get resolved with the SVG fix.

swissspidy commented on packages/design-system/src/components/datalist/list/list.tsx:
```diff
@@ -118,46 +119,50 @@ const OptionList = forwardRef(function OptionList(
-          // Gets ~2 extra items below scroll container BoundingBox
-          rootMargin: '60px',
+  const observer = useMemo(() => {
+    if (!listRef) {
```
Should this be

```suggestion
    if (!listRef.current) {
```

?

barklund replied:
Yes, it should be. Fixing.

Commit:
A bunch of feedback

Commit:
Merge branch 'main' into code/12659-ds-ts

Commit:
Trying to fix the remaining type issues including SVG issues

Commit:
Trying to fix more tests

Commit:
Fixing wrong component

Commit:
Trying to fix weird TS error happening in CI lint

Commit:
Fixing some last enum issues

Commit:
Merge branch 'main' into code/12659-ds-ts

Commit:
Trying to fix some errors

Commit:
Fixed a number of stale proptype references

Commit:
Merge branch 'main' into code/12659-ds-ts

swissspidy committed:
Use `import type` more often for type-only imports

swissspidy committed:
Remove duplicate icon

swissspidy committed:
Rename file to .ts

swissspidy committed:
Refactor Disclosure component to use styled svg icon directly

swissspidy committed:
Try to fix svg imports

swissspidy commented on packages/react/tsconfig.json:
```diff
@@ -2,7 +2,8 @@
     "rootDir": "src",
-    "declarationDir": "dist-types"
+    "declarationDir": "dist-types",
+    "baseUrl": "."
```
```suggestion
    "declarationDir": "dist-types"
```

Commit:
Removed irrelevant config

Commit:
Merge branch 'main' into code/12659-ds-ts

# Conflicts:
#	packages/design-system/src/components/circularProgress/circular.tsx
#	packages/design-system/src/components/contextMenu/styled.tsx
#	packages/design-system/src/components/dialog/dialog.tsx
#	packages/design-system/src/components/snackbar/snackbarContainer.tsx
#	packages/design-system/src/components/snackbar/snackbarMessage.tsx
#	packages/design-system/src/components/typography/display/index.ts
#	packages/design-system/src/components/typography/headline/index.ts
#	packages/design-system/src/components/typography/link/index.tsx
#	packages/design-system/src/components/typography/text/index.ts

Commit:
Simplified logic

Commit:
Fixed annoying popup bug!

barklund commented on packages/design-system/src/components/popup/popup.tsx:
```diff
@@ -85,50 +110,62 @@ function Popup({
 
-      const updatedXOffset = () => {
+      const updatedXOffset = (): Offset | null => {
+        if (!popupRect) {
+          return null;
+        }
```
Damn, this bug took me all day to find. Basically, the old code destructured an empty object into DOMRect attributes in case of the popup not showing yet and depended on `undefined < someLimit === NaN`, which is falsey. I used a default value of 0 in the destructuring to make TS happy, but that actually made the statement true, because suddenly it _was_ smaller than the limit value, which `undefined` _"isn't"_. So with the proper checks for the missing DOMRect just skipping this whole section of code, and everything works again.

Commit:
Merge branch 'main' into code/12659-ds-ts

Commit:
Fixed another bug relying on type error

Commit:
Fixed some styles and some StoryBook stories

Commit:
Fixed a state woopsi

Commit:
More precise fix

Commit:
Fixed some more storybook

Commit:
Fixed remaining storybooks

Commit:
Not sure why these are suddenly needed?

Commit:
Renamed some props because underlying component was refactored

Commit:
Fixed some more menu instances

Commit:
Fixed it for real

Commit:
Fixed util function (:mind-blown:)

Commit:
Fixed weird styling of focus trap button

Why is this even a primary button, when it does not follow the primary button styles? Could it not just be a secondary button?

Commit:
Beginning fix for grid view rearrange

Commit:
Fixing some more grid view bugs

Commit:
Fixed error with textarea without maxLength

Commit:
Merge branch 'main' into code/12659-ds-ts

Commit:
Merge branch 'main' into code/12659-ds-ts

Commit:
Fixed spelling of "Quaternary"

Commit:
Fixed checklist icon

Commit:
Fixed header title font size

Commit:
Fixed other font-size related issues

Commit:
Made grid view butten identical to other footer buttons

Commit:
Fixed some placement enum errors

Commit:
Fix some style overrides

Commit:
Updated button for hidden toolbar in footer to match the rest

Commit:
Fix bug when using arrow keys without selection

Commit:
Cleaned logic up a bit

Commit:
Adjusted footer menu alignment

Commit:
Merge branch 'main' into code/12659-ds-ts

# Conflicts:
#	packages/story-editor/src/components/floatingMenu/elements/shared/icon.js

Commit:
Fixed meta box button

Commit:
Fixed dashbaord font size

Commit:
Fixed link button in dialog

Commit:
Merge branch 'main' into code/12659-ds-ts

# Conflicts:
#	package-lock.json

Commit:
Fixed prop type

Commit:
Merge branch 'main' into code/12659-ds-ts

Commit:
Added no-flip to the transform

Commit:
Merge branch 'main' into code/12659-ds-ts

Commit:
Merge branch 'main' into code/12659-ds-ts

# Conflicts:
#	packages/story-editor/src/app/quickActions/useQuickActions.js
